lib/services/plugin-variables-service.ts

```diff
@@ -0,0 +1,102 @@
+					}
+					values[pluginVariableData.name] = pluginVariableValue;
+				}).future<void>()()).wait();
+
+			this.$projectDataService.initialize(this.$projectData.projectDir);
+			this.$projectDataService.setValue(this.getPluginVariablePropertyName(pluginData), values).wait();
+		}).future<void>()();
+	}
+
+	public removePluginVariablesFromProjectFile(pluginData: IPluginData): IFuture<void> {
+		this.$projectDataService.initialize(this.$projectData.projectDir);
+		return this.$projectDataService.removeProperty(this.getPluginVariablePropertyName(pluginData));
+	}
+
+	public interpolatePluginVariables(pluginData: IPluginData, pluginConfigurationFileContent: string): IFuture<string> {
+		return (() => {
+			this.executeForAllPluginVariables(pluginData, (pluginVariableData: IPluginVariableData) =>
+				(() => {
+					if(!pluginVariableData.value) {
+						this.$errors.failWithoutHelp(`Unable to find the value for ${pluginVariableData.name} plugin variable into project package.json file. Verify that your package.json file is correct and try again.`);
+					}
+
+					pluginConfigurationFileContent = pluginConfigurationFileContent.replace(new RegExp(`{${pluginVariableData.name}}`, "gi"), pluginVariableData.value);
+				}).future<void>()()).wait();
+			return pluginConfigurationFileContent;
+		}).future<string>()();
+	}
+
+	private getPluginVariableValue(pluginVariableData: IPluginVariableData): IFuture<string> {
+		return (() => {
+			let pluginVariableName = pluginVariableData.name;
+			let value = this.$pluginVariablesHelper.getPluginVariableFromVarOption(pluginVariableName);
+			if(value) {
+				value = value[pluginVariableName];
+			} else {
+				value = pluginVariableData.defaultValue;
+				if(!value && helpers.isInteractive()) {
+					let promptSchema = {
+						name: pluginVariableName,
+						type: "input",
+						message: `Enter value for ${pluginVariableName} variable:`,
+						validate: (val: string) => !!val ? true : 'Please enter a value!'
+					};
+					let promptData = this.$prompter.get([promptSchema]).wait();
+					value = promptData[pluginVariableName];
+				}
+			}
+
+			return value;
+		}).future<string>()();
+	}
+
+	private executeForAllPluginVariables(pluginData: IPluginData, action: (pluginVariableData: IPluginVariableData) => IFuture<void>): IFuture<void> {
+		return (() => {
+			let pluginVariables = pluginData.pluginVariables;
+			let pluginVariablesNames = _.keys(pluginVariables);
+			_.each(pluginVariablesNames, pluginVariableName => action(this.createPluginVariableData(pluginData, pluginVariableName).wait()).wait());
+		}).future<void>()();
+	}
+
+	private createPluginVariableData(pluginData: IPluginData, pluginVariableName: string): IFuture<IPluginVariableData> {
+		return (() => {
+			let variableData = pluginData.pluginVariables[pluginVariableName];
+
+			variableData.name = pluginVariableName;
+
+			this.$projectDataService.initialize(this.$projectData.projectDir);
```

Comment:
This is implementation detail. The code simply does not belong to the callback. In case we decide to refactor projectDataService  we need to remember to change this code as well.